File: DynamicPrograming/csharp/QuickestConcat.cs
using System;
using System.Collections.Generic;

namespace DynamicProgrammingSolutions;

public static class QuickestConcat
{
    public static int Quickest(string s, IReadOnlyList<string> words)
    {
        var memo = new Dictionary<int, int>();
        var result = Dfs(0, s, words, memo);
        return result >= int.MaxValue / 2 ? -1 : result;
    }

    private static int Dfs(int index, string s, IReadOnlyList<string> words, IDictionary<int, int> memo)
    {
        if (index == s.Length)
        {
            return 0;
        }

        if (memo.TryGetValue(index, out var cached))
        {
            return cached;
        }

        var best = int.MaxValue / 2;
        foreach (var word in words)
        {
            if (s.AsSpan(index).StartsWith(word))
            {
                best = Math.Min(best, 1 + Dfs(index + word.Length, s, words, memo));
            }
        }

        memo[index] = best;
        return best;
    }
}

File: DynamicPrograming/java/QuickestConcat.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class QuickestConcat {
  private QuickestConcat() {
  }

  public static int quickest(String s, List<String> words) {
    Map<Integer, Integer> memo = new HashMap<>();
    int result = dfs(0, s, words, memo);
    return result >= Integer.MAX_VALUE / 2 ? -1 : result;
  }

  private static int dfs(int index, String s, List<String> words, Map<Integer, Integer> memo) {
    if (index == s.length()) {
      return 0;
    }
    if (memo.containsKey(index)) {
      return memo.get(index);
    }
    int best = Integer.MAX_VALUE / 2;
    for (String word : words) {
      if (s.startsWith(word, index)) {
        best = Math.min(best, 1 + dfs(index + word.length(), s, words, memo));
      }
    }
    memo.put(index, best);
    return best;
  }
}
